### ISSUE TITLE
Add a getZoomBounds callback

### ISSUE THREAD

**Issue by brendanheywood** [domoritz/leaflet-locatecontrol#218]
At the moment the options for setting the zoom level are based purely on the accuracy of the location and not on anything else.

Some use cases:

1) the 'directions' use case: ie I'm in the google maps app and driving somewhere getting turn by turn directions. If you are on the open road you get zoomed out and follow at a high zoom level. But as you approach a feature like a turn or roundabout then the map zooms in, and then zooms out again after the turn.

2) the use case I have is similar, as I am navigating into and out of a nested set of boundaries for polygon features. Depending on where I am in this nested hierarchy I want to show a zoom level which is broadly the same size as that feature.

Suggested solution:

```
getZoomBounds: function(map, latlng, accuracy) {
    ... figure out bounds
    return bounds;
}
```

I can also imagine a use case where that call is async so maybe if you return null then it just updates the pan and it can do whatever it wants later eg:

```
getZoomBounds: function(map, latlng, accuracy) {
    someAjax.then(function() {
        map.fitBounds(bounds);
    });
    return null;
}
```

I think the result of the above would be used exactly and all the other options like maxZoom, circlePadding etc would be ignored as you could build that into your custom function. It would still honor the flyTo vs panTo

**Comment by domoritz** [comment]
`keepCurrentZoomLevel` addresses part of this but I can see the value of more flexibility. I do want to keep the library simple, though.

**Comment by brendanheywood** [comment]
This should almost be a 1 liner here:

https://github.com/domoritz/leaflet-locatecontrol/blob/gh-pages/src/L.Control.Locate.js#L485

If getZoomBounds is set then call is, otherwise use the event bounds. I'll make a pr so you can judge it properly

**Comment by domoritz** [comment]
Wouldn't it be easier for the user to return a zoom level? Can you think of a way to let people do that?

**Comment by brendanheywood** [comment]
I did think of that but no I think it's almost always better to return a bounds so it works at various viewport sizes, mobile vs desktop where the same bounds ends up as different zoom levels. Like you generally want to say things like 'make sure this feature is in the viewport' rather than calculate a zoom level based on the current viewport size in order to make sure than the feature ends up contained within it. This means you can also handle cases where you want to track the location but it isn't centered in the viewport, like you could have a consistent offset so you can see ahead if it was driving directions. Or say if you are inside a uni campus then you could always make the uni map fill the viewport and it wouldn't move at all, just the location within it. If you are outside the campus then you could make the returned bounds include both the gps location and the campus so you can see the directions to it. All the current code deals with bounds so doing it this way also makes the patch much simpler and pushes the complexity, and the flexibility, into the calling code where you want it.
